File: bin/gpt.js
#!/usr/bin/env node

const https = require('https');
const querystring = require('querystring');
const readline = require('readline');
const { URL } = require('url');

function answer(message, system = null, url = null, model = null) {
    url = url || process.env.GPT_URL;
    if (!url) {
        throw new Error("No GPT URL provided! Please use --url or set GPT_URL environment variable.");
    }

    const params = { message: message };
    if (system) params.system = system;
    if (model) params.model = model;

    const queryString = querystring.stringify(params);
    const fullUrl = new URL(`${url}?${queryString}`);
    const options = {
        headers: {
            "User-Agent": "node-https/1.0",  // Bypass Cloudflare
        }
    };
    const parts = [];
    https.get(fullUrl, options, (response) => {
        const rl = readline.createInterface({
            input: response,
            terminal: false
        });

        if (response.statusCode !== 200) {
            console.error(`Request Status Code: ${response.statusCode}`);
        }
        rl.on('line', (line) => {
            if (response.statusCode !== 200) {
                console.error(`${line}`);
            } else {
                const stripped = line.replace('data: ', '').trim();

                if (stripped && !stripped.includes('[DONE]')) {
                    try {
                        const data = JSON.parse(stripped);
                        const part = data.response || '';
                        if (part === '.' && parts[parts.length - 1] === '.') {
                            return;
                        }
                        parts.push(part);
                        process.stdout.write(part);
                    } catch (e) {
                        console.error(`Could not parse: ${stripped}`, e);
                        throw new Error(`Could not parse: ${stripped}`);
                    }
                }
            }
        });
    }).on('error', (e) => {
        console.error(e);
    });
}

function main() {
    const args = process.argv.slice(2);
    let system = null;
    let url = null;
    let model = null;
    let messages = [];

    // Simple argument parsing
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--system') {
            system = args[++i];
        } else if (args[i] === '--url') {
            url = args[++i];
        } else if (args[i] === '--model') {
            model = args[++i];
        } else {
            messages.push(args[i]);
        }
    }
    if (messages.length === 0) {
        let data = [];
        process.stdin.on('readable', () => {
            let chunk;
            while ((chunk = process.stdin.read()) !== null) {
                data.push(chunk);
            }
        });
        process.stdin.on('end', () => {
            messages.push(data.join(''));
            answer(messages.join(''), system, url, model);
        });
    } else {
        answer(messages.join(' '), system, url, model);
    }
    // 
}

if (require.main === module) {
    main();
}

File: bin/gpt.py
#!/usr/bin/env python3

from urllib.request import urlopen, Request
from urllib.parse import urlencode
import json
import logging
import argparse
import os

def answer(message, system=None, url=None, model=None):
    logger = logging.getLogger(__name__)

    url = url or os.environ.get('GPT_URL', None)
    if not url:
        raise ValueError("No GPT URL provided! Please use --url or set GPT_URL environment variable.")

    params = {'message': message}
    if system:
        params['system'] = system
    if model:
        params['model'] = model

    query_string = urlencode(params)
    full_url = f"{url}?{query_string}"
    req = Request(full_url, headers={
       "User-Agent": "python-requests/2.26.0",  # Bypass Cloudflare
    })
    parts = []
    with urlopen(req) as response:
        # Stream the response
        for line in response:
            decoded = line.decode('utf-8')
            stripped = decoded.replace('data: ', '', 1).strip()
            if stripped and '[DONE]' not in stripped:
                try:
                    data = json.loads(stripped)
                    part = data.get('response', '')
                    if part == '.' and parts[-1] == '.':
                        continue
                    parts.append(part)
                    print(part, end='')
                except Exception as e:
                    logger.error(f'Could not parse: {stripped}', e)
                    raise Exception(f'Could not parse: {stripped}')
    return ''.join(parts)


def main():
    import sys

    parser = argparse.ArgumentParser(description="Ask request to GPT endpoint and print the answers.")
    parser.add_argument('message', nargs='*', help="Message to send or read from stdin")
    parser.add_argument('--system', help="Optional system prompt to send with the message")
    parser.add_argument('--url', help="Optional Worker URL or set GPT_URL environment variable")
    parser.add_argument('--model', help="Optional model to use. See CF models for available models.")
    args = parser.parse_args()

    if not args.message:
        args.message = sys.stdin.read()

    answer(' '.join(args.message), args.system, args.url, args.model)

if __name__ == "__main__":
    main()
